### ISSUE TITLE
0 allocated memory after several apply and release

### ISSUE THREAD

**Issue by shaowenzh** [memkind/memkind#216]
Hello,

I am new to memkind and memory. Recently I am running some test about memkind in order to enable our project on persistent memory. I mounted a tmpfs as the path for memkind.  I initialized a 32mb memory pool and trying to apply for memory. Basically I tried to apply 3 times. And the memory will be released immediately after applying. The first time is 20mb. Second is 1024 byte. Third one is 20mb again. But when I run` memkind_malloc_usable_size` on the third time, it shows 0. Is this because of the memory fragmentation? Or do I use it in the right way? Also if I change the second time to apply for 20mb as well. All the three time will have right applied size.
```
#include <memkind.h>
#include <stdio.h>

#define PMEM_MAX_SIZE (1024 * 1024 * 32)

static void print_err_message(int err)
{
    char error_message[MEMKIND_ERROR_MESSAGE_SIZE];
    memkind_error_message(err, error_message, MEMKIND_ERROR_MESSAGE_SIZE);
    fprintf(stderr, "%s\n", error_message);
}

int main(int argc, char *argv[])
{
    struct memkind *pmemkind = NULL;
    char path[1024] = "/home/zsw/aep/1/memkindtest/";

    int error = 0;
    error = memkind_create_pmem(path, PMEM_MAX_SIZE, &pmemkind);
    if (error){
      print_err_message(error);
      return 1;
    }
    size_t mem20mb = (size_t)1024 * 1024 * 20;

    void *pmem1 = memkind_malloc(pmemkind, mem20mb);
    size_t sp1 = memkind_malloc_usable_size(pmemkind, pmem1);
    fprintf(stdout, "20mb: %d\n",sp1);
    memkind_free(pmemkind, pmem1);

    void *pmem2 = memkind_malloc(pmemkind, 1024);
    size_t sp2 = memkind_malloc_usable_size(pmemkind, pmem2);
    fprintf(stdout, "1024: %d\n",sp2);
    memkind_free(pmemkind, pmem2);

    void *pmem3 = memkind_malloc(pmemkind, mem20mb);
    size_t sp3 = memkind_malloc_usable_size(pmemkind, pmem3);
    fprintf(stdout, "20mb: %d\n",sp3);
    memkind_free(pmemkind, pmem3);
}
```

Thanks!

**Comment by michalbiesek** [comment]
Hi,
Thanks for your message.
Let me answer to your questions.

In your case: 
**memkind_malloc_usable_size(pmemkind, pmem3)** returns 0, because allocation described by:
**void \*pmem3 = memkind_malloc(pmemkind, mem20mb)** fails, which results that void *pmem3 is NULL.

Memkind is a heap manager build on top of **jemalloc** which enables controly of memory characteristics.
Jemalloc, which is under memkind used two-phase decay of unused dirty pages, which results that your memory doesn't come back to OS immediately after free.
Memkind provides an option to manipulate this mechanism.
You could try create your pmem kind using **memkind_create_pmem_with_config()** with **memkind_mem_usage_policy** set as **MEMKIND_MEM_USAGE_POLICY_CONSERVATIVE**.

**Comment by shaowenzh** [comment]
@michalbiesek Thanks a lot for the quick response! I tried the way you mentioned above. It works! But for another scenario like this:
```
    struct memkind_config *pmemkind_config;
    pmemkind_config = memkind_config_new();
    memkind_config_set_path(pmemkind_config, path);
    memkind_config_set_size(pmemkind_config, PMEM_MAX_SIZE);
    memkind_config_set_memory_usage_policy(pmemkind_config, 
       MEMKIND_MEM_USAGE_POLICY_CONSERVATIVE);

    int error = 0;
    error = memkind_create_pmem_with_config(pmemkind_config, &pmemkind);

    void *pmem3 = memkind_malloc(pmemkind, 1024 * 1024 * 16);
    size_t sp3 = memkind_malloc_usable_size(pmemkind, pmem3);
    fprintf(stdout, "16mb: %d\n",sp3);
    memkind_free(pmemkind, pmem3);

    void *pmem4 = memkind_malloc(pmemkind, 1024 * 1024 * 17);
    size_t sp4 = memkind_malloc_usable_size(pmemkind, pmem3);
    fprintf(stdout, "17mb: %d\n",sp4);
    memkind_free(pmemkind, pmem4);
```
Still cache pool size is **32mb**. I tried to applied for memory two times. First is **16mb** and second is **17mb**. Total size is more than the pool's limitation.  But even when I immediately release the **first** applied memory. And the size for **second** pmem4 only got assigned **4096** byte size. Is this also because the dirty pages not come back to OS directly after free? 

Just would like to understand the memkind behavior right :). Our initial design by using memkind seems got lots of **allocation error** when the memory is full and we do use the `memkind_free` to try to free more memory.

**Comment by michalbiesek** [comment]
>     size_t sp4 = memkind_malloc_usable_size(pmemkind, pmem3)
I think it should be pmem4 not pmem3 :)

But scenario you present looks like an issue (allocation failure with pmem4) - this looks like a problem with jemalloc. We will focus on it in the future.

BTW can you share initial design in which you want use the memkind?

**Comment by shaowenzh** [comment]
@michalbiesek Oops, sorry for the typo here.  I changed it to pmem4, it shows 0 this time. 

We are using Google Guava's LRU mechanism to manage the key for the cache. We measure the total cache size by using the size which we are trying to applied from **memkind**. Once the total size exceed the LRU's maximum limitation, which we normally set it a little bit smaller than the value we call **memkind_create_pmem** , it will start to evict the cache that is not used recently and we will call **memkind_free** to free them. But later we figured out that this size will be smaller than the real allocated size which we could get from **memkind_malloc_usable_size**, I think next we will try to use the  **memkind_malloc_usable_size** to measure the cache size, maybe this will give a better performance. But for the scenario I mentioned above, I don't know how much it will impact the result yet. And for the cache space application and free, they will happen under multi-thread condition.

Thanks for the time to answer my question!
